sharktank/sharktank/export.py

```diff
@@ -0,0 +1,156 @@
+        [ReplicatedTensor(ts=[torch.tensor([2]), torch.tensor([3])])]
+    )
+    ```
+    returns
+    ```
+    {
+        1: DeviceAffinity("0"),
+        2: DeviceAffinity("1"),
+    }
+    ```
+    """
+
+    def is_leaf(v: PyTree) -> bool:
+        if isinstance(v, ShardedTensor):
+            return True
+        # TODO: It is sad _is_leaf is private. Find a way not use it.
+        from torch.utils._pytree import _is_leaf
+
+        return _is_leaf(v)
+
+    # flattened up to a sharded tensor.
+    flat_args_up_to_sharded_tensor = tree_flatten(args, is_leaf=is_leaf)[0]
+    nested_device_affinities: list[list[DeviceAffinity | None]] = [
+        [DeviceAffinity(f"{shard_idx}") for shard_idx in range(len(arg.shards))]
+        if isinstance(arg, ShardedTensor)
+        else [None]
+        for arg in flat_args_up_to_sharded_tensor
+    ]
+    flat_device_affinities: list[DeviceAffinity | None] = [
+        affinity
+        for affinity_list in nested_device_affinities
+        for affinity in affinity_list
+    ]
+    return {
+        arg_idx: affinity
+        for arg_idx, affinity in enumerate(flat_device_affinities)
+        if affinity is not None
+    }
+
+
+def export(
+    f: Callable,
+    /,
+    fx_builder: FxProgramsBuilder | None = None,
+    args: tuple[PyTree] | None = None,
+    argument_device_affinities: dict[int, DeviceAffinity] | None = None,
+    *transitive_args,
+    **transitive_kwargs,
+) -> torch.export.ExportedProgram:
+    """Wrapper around FxProgramsBuilder.export_program that handles
+    the sharktank custom tensor types.
+
+    If `argument_device_affinities` is not specified it will extract the affinities
+    from the passed `args`.
+    `argument_device_affinities` must pass the affinities for the flattened arguments.
+    These are those that correspond to torch.Tensor.
+    For example a sharded tensor with 2 shards would result in 2 arguments in the MLIR
+    signature."""
+    if args is None:
+        args = []
+    if argument_device_affinities is None:
+        argument_device_affinities = get_argument_flat_device_affinities(*args)
+    flat_args = tree_flatten(args)[0]
+    if fx_builder is not None:
+        # Flatten the signature of the function.
+        # Technically this is done during export, but we want the signature to match
+        # the flat device affinities.
+        def module_fn_with_flat_signature(module, *flat_args):
+            @flatten_signature(*args)
+            def flat_fn(*args):
+                return f(module, *args)
+
+            return flat_fn(*flat_args)
+
+        amended_kwargs = dict(**transitive_kwargs)
+        if "name" not in amended_kwargs or amended_kwargs["name"] is None:
+            amended_kwargs["name"] = f.__name__
+        return fx_builder.export_program(
+            module_fn_with_flat_signature,
+            *transitive_args,
+            args=flat_args,
+            argument_device_affinities=argument_device_affinities,
```

Comment:
Wait, where does this actually plumb through? I don't see this as part of the PR

Reply:
I see, the changes are dependent on https://github.com/iree-org/iree-turbine/pull/231/files
Lets focus on that and shelve these changes. Once `export` has the right parameter we won't need to make the sharktank changes.